Multitool Upgrade

PsiOmegaDelta opened the pull request:
This is a new system for expanding objects with new behaviors in a component based manner.
This is done using a list on obj, named expansions and can be initialized as follows:
expansions[/datum/expansion/type] = new/datum/expansion/type()
Practical example [here](https://github.com/Baystation12/Baystation12/pull/11458/files#diff-1e264196f81ead51bb747ac7d38f1a09R60).

The multitool uses this method to allow objects to have specific multitool menus without having to modify the original type definition.
Included in this PR is the ability to:
- Connect/disconnect cloning pods from cloning computers.
- Toggle enabled station alerts on a station alert console circuit, which then affects a finished computer console.

PsiOmegaDelta committed:
Multitool menu.

Begins work on allowing the multitool to modify and alter more equipment than ever.

PsiOmegaDelta committed:
Can now connect cryo pods and computers with each other using the new multitool methodology.

Also cleans up Destroys().

PsiOmegaDelta committed:
Allows the AI to utilize its multitool where applicable.

PsiOmegaDelta committed:
It is now possible to use a multitool to configure alert console circuits, and the available alarm sources.

PsiOmegaDelta committed:
Merge remote-tracking branch 'upstream/dev' into 151117-MultitoolPower

Conflicts:
	code/game/objects/items/devices/multitool.dm

PsiOmegaDelta committed:
The multitool now properly releases its buffer reference when the buffer is destroyed.

PsiOmegaDelta committed:
Changelog entry.

PsiOmegaDelta committed:
Merge remote-tracking branch 'upstream/dev' into 151117-MultitoolPower

Conflicts:
	code/modules/mob/living/silicon/ai/ai.dm

PsiOmegaDelta committed:
Merge remote-tracking branch 'upstream/dev' into 151117-MultitoolPower

PsiOmegaDelta commented:
Fixes #11569.

PsiOmegaDelta committed:
Merge remote-tracking branch 'upstream/dev' into 151117-MultitoolPower

PsiOmegaDelta committed:
Merge remote-tracking branch 'upstream/dev' into 151117-MultitoolPower

Conflicts:
	code/game/machinery/computer/atmos_alert.dm
	code/game/machinery/computer/station_alert.dm

Ccomp5950 commented:
Looking over this, code looks good.  I didn't spin up a server and play with it yet which I'll do later today (at work for another hour).

PsiOmegaDelta committed:
Fixes #11569.

mwerezak commented:
A bit of a UI layout issue on the alert board configuration page. Didn't notice any other issues with either of the currently supported configurable objs.

![screenshot 2015-12-05 13 42 49](https://cloud.githubusercontent.com/assets/242428/11609513/419dadae-9b56-11e5-8c88-dca28d59a73f.png)

PsiOmegaDelta commented:
I can make the window excessively wide to cover for most potential title widths, or add a number of linebreaks to ensure the title rarely overlaps.

PsiOmegaDelta committed:
Extension additions and multitool menu tweaks.

Relocates obj extension code to the extension file.
Adds a support method, set_extension, to ensure that proper cleanup occurs in the case of multiple additions to the same extension type.
The default multitool menu title is now much shorter.
The cloning computer multitool menu now lists all connected cloning pods.

PsiOmegaDelta commented:
Relocates obj extension code to the extension file.
Adds a support method, set_extension, to ensure that proper cleanup occurs in the case of multiple additions to the same extension type.
The default multitool menu title is now much shorter.
The cloning computer multitool menu now lists all connected cloning pods.

PsiOmegaDelta committed:
Merge remote-tracking branch 'upstream/dev' into 151117-MultitoolPower

Conflicts:
	code/modules/nano/nanoui.dm

PsiOmegaDelta committed:
Merge remote-tracking branch 'upstream/dev' into 151117-MultitoolPower
